pkg/proxy/iptables/proxier.go

```diff
@@ -1073,58 +1019,121 @@ func (proxier *Proxier) syncProxyRules() {
+						"-j", string(KubeMarkMasqChain))
+				}
+				// Likewise all traffic to a to ClusterIP from a non-pod
+				// source gets masqueraded (if we know how to distinguish
+				// pods from non-pods).
+				if !svcInfo.NodeLocalInternal() && proxier.localDetector.IsImplemented() {
+					proxier.natRules.Write(
+						"-A", string(svcChain),
+						"-d", svcInfo.ClusterIP().String(),
+						proxier.localDetector.IfNotLocal(),
+						"-j", string(KubeMarkMasqChain))
+				}
+			}
+		}
+		internalTrafficChain := svcChain
+		loadBalancerTrafficChain := svcChain
+		nodePortTrafficChain := svcChain
+
+		svcLocalChain := svcInfo.serviceLocalChainName
+		if hasEndpoints && (svcInfo.NodeLocalInternal() || svcInfo.NodeLocalExternal()) {
+			// create the Local traffic policy chain, retaining counters if possible.
+			if chain, ok := existingNATChains[svcLocalChain]; ok {
+				proxier.natChains.WriteBytes(chain)
+			} else {
+				proxier.natChains.Write(utiliptables.MakeChainLine(svcLocalChain))
+			}
+			activeNATChains[svcLocalChain] = true
 		}
 
 		svcXlbChain := svcInfo.serviceLBChainName
 		if hasEndpoints && svcInfo.NodeLocalExternal() {
-			// Only for services request OnlyLocal traffic
-			// create the per-service LB chain, retaining counters if possible.
-			if lbChain, ok := existingNATChains[svcXlbChain]; ok {
-				proxier.natChains.WriteBytes(lbChain)
+			// create the XLB short-circuiting chain, retaining counters if possible
+			if chain, ok := existingNATChains[svcXlbChain]; ok {
+				proxier.natChains.WriteBytes(chain)
 			} else {
 				proxier.natChains.Write(utiliptables.MakeChainLine(svcXlbChain))
 			}
 			activeNATChains[svcXlbChain] = true
+
+			// The XLB chain redirects all pod -> external VIP
+			// traffic to the Service's ClusterIP instead. This happens
+			// whether or not we have local endpoints; only if localDetector
+			// is implemented
+			if proxier.localDetector.IsImplemented() {
+				proxier.natRules.Write(
+					"-A", string(svcXlbChain),
+					"-m", "comment", "--comment",
+					`"Redirect pods trying to reach external loadbalancer VIP to clusterIP"`,
+					proxier.localDetector.IfLocal(),
+					"-j", string(svcChain))
+			}
+
+			// Next, redirect all src-type=LOCAL -> LB IP to the service chain
+			// for externalTrafficPolicy=Local This allows traffic originating
+			// from the host to be redirected to the service correctly,
+			// otherwise traffic to LB IPs are dropped if there are no local
+			// endpoints.
+			proxier.natRules.Write(
+				"-A", string(svcXlbChain),
+				"-m", "comment", "--comment", fmt.Sprintf(`"route LOCAL traffic for %s LB IP to service chain"`, svcNameString),
+				"-m", "addrtype", "--src-type", "LOCAL",
+				"-j", string(svcChain))
+
+			// Everything else goes to the SVL chain
+			proxier.natRules.Write(
+				"-A", string(svcXlbChain),
+				"-j", string(svcLocalChain))
+		}
+
+		if svcInfo.NodeLocalInternal() {
+			internalTrafficChain = svcLocalChain
```

Comment:
The comments for serviceLocalChainName (KUBE-SVL) mentions it is only used for external traffic policy

Reply:
updated that. (It still says external-only in the original commit, but gets updated in the final commit)